### PR TITLE
SCM-798 feat(package): add support for custom formats

### PR DESCRIPTION
**Pull request by michaelcarter-wf**
## Ultimate problem:
- We only partially support custom formats (redefining validation for existing formats)
- Our release notes aren't quite ready for 4.0

## How it was fixed:
- Allow custom formats in a similar way we expose custom vocabularies
- Deprecate `DefaultValidators` and the global getter / setter.
- Dogfood the basic API for our own format validators.
- Break out into 1-format-validator-per-file
- Reorganize some other directories
- Update release notes

## Potential areas of regression:

---

> __FYA:__ @michaelcarter-wf

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: example/custom_format/custom_format_example.dart [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+//     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//     THE SOFTWARE.
+
+import 'package:json_schema/json_schema.dart';
+import 'package:json_schema/src/json_schema/models/validation_context.dart';
+
+ValidationContext screamingCapsValidator(ValidationContext context, String instanceData) {
+  if (instanceData.toUpperCase() != instanceData) {
+    context.addError('"screaming-caps" format not accepted $instanceData');
```
Comment (tomdeering-wf) [inline example/custom_format/custom_format_example.dart:44]: LOL :1st_place_medal: for this example

File: lib/src/json_schema/formats/hostname_validator.dart [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import 'package:json_schema/src/json_schema/formats/validation_regexes.dart';
+import 'package:json_schema/src/json_schema/models/schema_version.dart';
+import 'package:json_schema/src/json_schema/models/validation_context.dart';
+
+ValidationContext defaultHostnameValidator(ValidationContext context, String instanceData) {
+  final regexp = context.schemaVersion > SchemaVersion.draft2019_09
+      ? JsonSchemaValidationRegexes.hostname
+      // Updated in Draft 2019-09
+      : JsonSchemaValidationRegexes.hostnameDraft2019;
```
Comment (tomdeering-wf) [inline lib/src/json_schema/formats/hostname_validator.dart:9]: Are these flipped?

File: lib/src/json_schema/formats/hostname_validator.dart [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import 'package:json_schema/src/json_schema/formats/validation_regexes.dart';
+import 'package:json_schema/src/json_schema/models/schema_version.dart';
+import 'package:json_schema/src/json_schema/models/validation_context.dart';
+
+ValidationContext defaultHostnameValidator(ValidationContext context, String instanceData) {
+  final regexp = context.schemaVersion > SchemaVersion.draft2019_09
```
Comment (tomdeering-wf) [inline lib/src/json_schema/formats/hostname_validator.dart:6]:
```suggestion
  final regexp = context.schemaVersion < SchemaVersion.draft2019_09
```

File: lib/src/json_schema/validator.dart [hunk 4]
```diff
@@ -550,158 +548,14 @@ class Validator {
...
-        final isValid = defaultValidators.uriReferenceValidator ?? (_) => false;
-
-        if (!isValid(instance.data)) {
-          _err('"uri-reference" format not accepted $instance', instance.path, schema.path);
-        }
-        break;
-      case 'uri-template':
-        if (![SchemaVersion.draft6, SchemaVersion.draft7].contains(schema.schemaVersion)) return;
-        final isValid = defaultValidators.uriTemplateValidator ?? (_) => false;
+    final validator = schema.customFormats[schema.format] ?? defaultFormatValidators[schema.format];
```
Comment (tomdeering-wf) [inline lib/src/json_schema/validator.dart:551]: The robbie pattern + defaults! I like it!

**Review by tomdeering-wf — CHANGES_REQUESTED** [review]
Otherwise looks great

**Comment by michaelcarter-wf** [comment]
+10 

- [x] CI Passes
- [x] Tests added to cover custom formats

**Comment by michaelcarter-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**